{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from glob import glob\n",
    "import subprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# outpath='../../results/FindMotifs/quantseq_groups'\n",
    "outpath='../../results/FindMotifs/full_utrs'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "patterns='AAA AATAAA GAT TTT ATT GGA,GGT'\n",
    "# outpath=../../results/FindMotifs/quantseq_groups\n",
    "outpath=../../results/FindMotifs/full_utrs\n",
    "\n",
    "\n",
    "# utrs=../../results/GeneGroups_Quantseq/3UtrLoci/QuantseqGroups.bed\n",
    "utrs=../../data/3UtrAtlas/ThreePrimeUtrsOfMostExpressedTxInS200WT2iL.bed\n",
    "genome=../../data/genomes/Goodwright_m39/GRCm39.primary_assembly.genome.fa\n",
    "\n",
    "\n",
    "mkdir -p $outpath\n",
    "\n",
    "# Extract 3'UTR fasta - ignore strand\n",
    "bedtools getfasta -name -fi $genome -bed $utrs > ${outpath}/full_utrs_noRevComp.fa\n",
    "\n",
    "# Connect headers to relevant strand\n",
    "awk '{new_var=$4\"::\"$1\":\"$2\"-\"$3; print new_var\"\\t\"$6}' $utrs > ${outpath}/id_strand.tsv\n",
    "\n",
    "\n",
    "# Locate subsequences - both positive and negative strand\n",
    "for p in $patterns\n",
    "do\n",
    "  # Replace comma with space to pass multiple strings\n",
    "  motifs=${p//','/' -p '}\n",
    "  cat ${outpath}/full_utrs_noRevComp.fa \\\n",
    "  | seqkit locate --ignore-case --max-mismatch 0 --hide-matched -p $motifs --bed \\\n",
    "    > ${outpath}/full_utrs_noRevComp_locate_${p}.tsv\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_ids = pd.read_csv(f'{outpath}/id_strand.tsv', sep='\\t', header=None, names=['id', 'strand']).astype(str)\n",
    "files = glob(f'{outpath}/full_utrs_noRevComp_locate_*.tsv')\n",
    "\n",
    "for f in files:\n",
    "    m = f.split('_')[-1].split('.')[0]\n",
    "    df_motifs = pd.read_csv(f, sep='\\t', names=['id', 'start', 'end', 'motif', 'score', 'strand'])\n",
    "    df_motifs[['id', 'strand']] = df_motifs[['id', 'strand']].astype(str)\n",
    "\n",
    "\n",
    "    # Use the inner merge to only keep identified sites for the relevant strand\n",
    "    df_motifs = df_ids.merge(df_motifs, on=['id', 'strand'], how='inner')\n",
    "\n",
    "    # Separate id elements to get genomic coordinates\n",
    "    df_motifs['reg_start'] = df_motifs.id.apply(lambda x: x.split(':')[-1].split('-')[0]).astype(int)\n",
    "    df_motifs['reg_end'] = df_motifs.id.apply(lambda x: x.split(':')[-1].split('-')[1]).astype(int)\n",
    "    df_motifs['chr'] = df_motifs.id.apply(lambda x: x.split(':')[2])\n",
    "    df_motifs['regulation'] = df_motifs.id.apply(lambda x: x.split('::')[0])\n",
    "\n",
    "    # Get genomic coordinates of motifs\n",
    "    df_motifs['mot_start'] = df_motifs[['reg_start', 'start']].sum(axis='columns')\n",
    "    df_motifs['mot_end'] = df_motifs[['reg_start', 'end']].sum(axis='columns')\n",
    "\n",
    "    # Save bed file - checked in IGV -  worked perfectly\n",
    "    bed = df_motifs[['chr', 'mot_start', 'mot_end', 'regulation', 'score', 'strand']]\n",
    "    bed.to_csv(f'{outpath}/motifs_{m}.bed.gz', sep='\\t', header=None, index=False, quoting=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "motifs='AAA AATAAA GAT TTT ATT GGA,GGT'\n",
    "# outpath=../../results/FindMotifs/quantseq_groups\n",
    "outpath=../../results/FindMotifs/full_utrs\n",
    "\n",
    "for m in $motifs; do\n",
    "    zcat $outpath/motifs_${m}.bed.gz | sort -k1,1 -k2,2n | bedtools merge -s -c 4,5,6 -o distinct,sum,distinct | gzip > $outpath/motifs_${m}_merged.bed.gz\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>strand</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>motif</th>\n",
       "      <th>score</th>\n",
       "      <th>reg_start</th>\n",
       "      <th>reg_end</th>\n",
       "      <th>chr</th>\n",
       "      <th>regulation</th>\n",
       "      <th>mot_start</th>\n",
       "      <th>mot_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000063316::chr11:101336310-101336355</td>\n",
       "      <td>+</td>\n",
       "      <td>25</td>\n",
       "      <td>28</td>\n",
       "      <td>AAA</td>\n",
       "      <td>0</td>\n",
       "      <td>101336310</td>\n",
       "      <td>101336355</td>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSMUSG00000063316</td>\n",
       "      <td>101336335</td>\n",
       "      <td>101336338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000063316::chr11:101336310-101336355</td>\n",
       "      <td>+</td>\n",
       "      <td>26</td>\n",
       "      <td>29</td>\n",
       "      <td>AAA</td>\n",
       "      <td>0</td>\n",
       "      <td>101336310</td>\n",
       "      <td>101336355</td>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSMUSG00000063316</td>\n",
       "      <td>101336336</td>\n",
       "      <td>101336339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000063316::chr11:101336310-101336355</td>\n",
       "      <td>+</td>\n",
       "      <td>27</td>\n",
       "      <td>30</td>\n",
       "      <td>AAA</td>\n",
       "      <td>0</td>\n",
       "      <td>101336310</td>\n",
       "      <td>101336355</td>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSMUSG00000063316</td>\n",
       "      <td>101336337</td>\n",
       "      <td>101336340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000063316::chr11:101336310-101336355</td>\n",
       "      <td>+</td>\n",
       "      <td>32</td>\n",
       "      <td>35</td>\n",
       "      <td>AAA</td>\n",
       "      <td>0</td>\n",
       "      <td>101336310</td>\n",
       "      <td>101336355</td>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSMUSG00000063316</td>\n",
       "      <td>101336342</td>\n",
       "      <td>101336345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000063316::chr11:101336310-101336355</td>\n",
       "      <td>+</td>\n",
       "      <td>33</td>\n",
       "      <td>36</td>\n",
       "      <td>AAA</td>\n",
       "      <td>0</td>\n",
       "      <td>101336310</td>\n",
       "      <td>101336355</td>\n",
       "      <td>chr11</td>\n",
       "      <td>ENSMUSG00000063316</td>\n",
       "      <td>101336343</td>\n",
       "      <td>101336346</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              id strand  start  end motif  \\\n",
       "0  ENSMUSG00000063316::chr11:101336310-101336355      +     25   28   AAA   \n",
       "1  ENSMUSG00000063316::chr11:101336310-101336355      +     26   29   AAA   \n",
       "2  ENSMUSG00000063316::chr11:101336310-101336355      +     27   30   AAA   \n",
       "3  ENSMUSG00000063316::chr11:101336310-101336355      +     32   35   AAA   \n",
       "4  ENSMUSG00000063316::chr11:101336310-101336355      +     33   36   AAA   \n",
       "\n",
       "   score  reg_start    reg_end    chr          regulation  mot_start  \\\n",
       "0      0  101336310  101336355  chr11  ENSMUSG00000063316  101336335   \n",
       "1      0  101336310  101336355  chr11  ENSMUSG00000063316  101336336   \n",
       "2      0  101336310  101336355  chr11  ENSMUSG00000063316  101336337   \n",
       "3      0  101336310  101336355  chr11  ENSMUSG00000063316  101336342   \n",
       "4      0  101336310  101336355  chr11  ENSMUSG00000063316  101336343   \n",
       "\n",
       "     mot_end  \n",
       "0  101336338  \n",
       "1  101336339  \n",
       "2  101336340  \n",
       "3  101336345  \n",
       "4  101336346  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_motifs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>mot_start</th>\n",
       "      <th>mot_end</th>\n",
       "      <th>regulation</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11</td>\n",
       "      <td>101336335</td>\n",
       "      <td>101336338</td>\n",
       "      <td>ENSMUSG00000063316</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11</td>\n",
       "      <td>101336336</td>\n",
       "      <td>101336339</td>\n",
       "      <td>ENSMUSG00000063316</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>101336337</td>\n",
       "      <td>101336340</td>\n",
       "      <td>ENSMUSG00000063316</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11</td>\n",
       "      <td>101336342</td>\n",
       "      <td>101336345</td>\n",
       "      <td>ENSMUSG00000063316</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr11</td>\n",
       "      <td>101336343</td>\n",
       "      <td>101336346</td>\n",
       "      <td>ENSMUSG00000063316</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr  mot_start    mot_end          regulation  score strand\n",
       "0  chr11  101336335  101336338  ENSMUSG00000063316      0      +\n",
       "1  chr11  101336336  101336339  ENSMUSG00000063316      0      +\n",
       "2  chr11  101336337  101336340  ENSMUSG00000063316      0      +\n",
       "3  chr11  101336342  101336345  ENSMUSG00000063316      0      +\n",
       "4  chr11  101336343  101336346  ENSMUSG00000063316      0      +"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bed.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.11.0 ('seqkit')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "b1aa3e4794e339171efe168fe1aa27ee014b1e186b9dea8ade0205aaefef340c"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
