{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pybedtools as pbt\n",
    "import pandas as pd\n",
    "from glob import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def readBed(f):\n",
    "    return pd.read_csv(f, sep='\\t', header=None)\n",
    "\n",
    "def GetStarts(df):\n",
    "    df.loc[df[5] == '+', 2] = df.loc[df[5] == '+', 1] + 1\n",
    "    df.loc[df[5] == '-', 1] = df.loc[df[5] == '-', 2] - 1\n",
    "    return df\n",
    "\n",
    "def StrandedSlop(bt, fai, UpstreamW=0, DownstreamW=300):\n",
    "    bt = bt.slop(s=True, l=UpstreamW, r=DownstreamW, g=fai).sort()\n",
    "    return bt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Files\n",
    "\n",
    "# Params\n",
    "# Region around PAS\n",
    "downW = 50\n",
    "upW = 200\n",
    "\n",
    "# Min 3'UTR length\n",
    "minLen = 300\n",
    "\n",
    "# terminal PAS in full 3'UTR set\n",
    "# terminal PAS gene-groups\n",
    "files = [\n",
    "    # '../../../data/TerminalPas/terminal_pa_signals.bed.gz',\n",
    "    '../../../data/TerminalPas/terminal_pa_signals_UP.bed.gz',\n",
    "    '../../../data/TerminalPas/terminal_pa_signals_CONTROL.bed.gz',\n",
    "    '../../../data/TerminalPas/terminal_pa_signals_DOWN.bed.gz',\n",
    "]\n",
    "# Quantseq regs\n",
    "dfUtrs = readBed('../../../results/GeneGroups_Quantseq/3UtrLoci/QuantseqGroups.bed')\n",
    "dfUtrs['len'] = dfUtrs[2] - dfUtrs[1]\n",
    "dfUtrs = dfUtrs.loc[dfUtrs.len >= minLen, :6]\n",
    "genesToPlot = dfUtrs[6].values.tolist()\n",
    "\n",
    "# fasta index\n",
    "fai = '../../../data/genomes/Goodwright_m39/GRCm39.primary_assembly.genome.fa.fai'\n",
    "\n",
    "# Expression\n",
    "DfTpm = pd.read_csv('../../../data/MihaDeseq/GeneLevel_TPM_counts.csv')\n",
    "\n",
    "# Naive genes\n",
    "naive_ids = pd.read_csv('../../../data/general/NaiveGeneIds.csv', index_col=0)\n",
    "\n",
    "# Save to\n",
    "out_pas_starts = '../../../data/TerminalPas/Starts'\n",
    "out_regions = 'ProfileRegions'\n",
    "out_combined_regions = '../../../results/Metaprofiles/PAS_Quantseq'\n",
    "os.makedirs(out_pas_starts, exist_ok=True)\n",
    "os.makedirs(out_regions, exist_ok=True)\n",
    "os.makedirs(out_combined_regions, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>KO_2iL Mean TPM</th>\n",
       "      <th>KO_FCL Mean TPM</th>\n",
       "      <th>S200A_2iL Mean TPM</th>\n",
       "      <th>S200A_FCL Mean TPM</th>\n",
       "      <th>S200WT_2iL Mean TPM</th>\n",
       "      <th>S200WT_FCL Mean TPM</th>\n",
       "      <th>WT_2iL Mean TPM</th>\n",
       "      <th>WT_FCL Mean TPM</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>stable_gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000001</th>\n",
       "      <td>7.677950</td>\n",
       "      <td>8.112323</td>\n",
       "      <td>6.880970</td>\n",
       "      <td>6.033886</td>\n",
       "      <td>8.122433</td>\n",
       "      <td>2.797935</td>\n",
       "      <td>4.333492</td>\n",
       "      <td>3.864102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000003</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000028</th>\n",
       "      <td>11.692164</td>\n",
       "      <td>9.045459</td>\n",
       "      <td>10.775650</td>\n",
       "      <td>7.133840</td>\n",
       "      <td>14.348023</td>\n",
       "      <td>4.322841</td>\n",
       "      <td>15.107630</td>\n",
       "      <td>7.196186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000031</th>\n",
       "      <td>1.093472</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.062915</td>\n",
       "      <td>1.569452</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.178640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000037</th>\n",
       "      <td>1.144660</td>\n",
       "      <td>1.038239</td>\n",
       "      <td>1.305144</td>\n",
       "      <td>1.137961</td>\n",
       "      <td>1.196149</td>\n",
       "      <td>1.249172</td>\n",
       "      <td>1.057581</td>\n",
       "      <td>1.086607</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    KO_2iL Mean TPM  KO_FCL Mean TPM  S200A_2iL Mean TPM  \\\n",
       "stable_gene_id                                                             \n",
       "ENSMUSG00000000001         7.677950         8.112323            6.880970   \n",
       "ENSMUSG00000000003         1.000000         1.000000            1.000000   \n",
       "ENSMUSG00000000028        11.692164         9.045459           10.775650   \n",
       "ENSMUSG00000000031         1.093472         1.000000            1.062915   \n",
       "ENSMUSG00000000037         1.144660         1.038239            1.305144   \n",
       "\n",
       "                    S200A_FCL Mean TPM  S200WT_2iL Mean TPM  \\\n",
       "stable_gene_id                                                \n",
       "ENSMUSG00000000001            6.033886             8.122433   \n",
       "ENSMUSG00000000003            1.000000             1.000000   \n",
       "ENSMUSG00000000028            7.133840            14.348023   \n",
       "ENSMUSG00000000031            1.569452             1.000000   \n",
       "ENSMUSG00000000037            1.137961             1.196149   \n",
       "\n",
       "                    S200WT_FCL Mean TPM  WT_2iL Mean TPM  WT_FCL Mean TPM  \n",
       "stable_gene_id                                                             \n",
       "ENSMUSG00000000001             2.797935         4.333492         3.864102  \n",
       "ENSMUSG00000000003             1.000000         1.000000         1.000000  \n",
       "ENSMUSG00000000028             4.322841        15.107630         7.196186  \n",
       "ENSMUSG00000000031             1.000000         1.000000         1.178640  \n",
       "ENSMUSG00000000037             1.249172         1.057581         1.086607  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get mean TPMs in each condition for gene-level TPMs\n",
    "DfTpm.set_index('stable_gene_id', drop=True, inplace=True)\n",
    "# To avoid 0-division error assign a +1 to all values\n",
    "DfTpm = DfTpm + 1\n",
    "DfTpm.head()\n",
    "conditions = []\n",
    "colnames = []\n",
    "for c in DfTpm.columns:\n",
    "    cond = '_'.join([c.split('_')[el] for el in [0, 2]])\n",
    "    colnames.append(f'{cond}.{c}')\n",
    "    conditions.append(cond)\n",
    "conditions = sorted(set(conditions))\n",
    "DfTpm.columns = colnames\n",
    "\n",
    "DfTpmAveraged = pd.DataFrame()\n",
    "for cond in conditions:\n",
    "    DfTpmAveraged[f'{cond} Mean TPM'] = DfTpm[[c for c in DfTpm.columns if c.split('.')[0] == cond]].mean(axis='columns')\n",
    "DfTpmAveraged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "terminal_pa_signals_UP 633\n",
      "terminal_pa_signals_CONTROL 1739\n",
      "terminal_pa_signals_DOWN 831\n"
     ]
    }
   ],
   "source": [
    "cond_of_interest = ['KO_FCL', 'S200WT_FCL', 'S200A_FCL', 'S200WT_2iL']\n",
    "for f in files:\n",
    "    btStarts = pbt.BedTool().from_dataframe(GetStarts(readBed(f))).sort()\n",
    "    fname = f.split('/')[-1].split('.')[0]\n",
    "    btStarts.saveas(f'{out_pas_starts}/{fname}.bed.gz')\n",
    "    dfRegs = StrandedSlop(btStarts, fai=fai, UpstreamW=upW, DownstreamW=downW).to_dataframe(disable_auto_names=True, header=None)\n",
    "    dfRegs = dfRegs.loc[dfRegs[3].isin(genesToPlot)]\n",
    "    print(fname, len(dfRegs))\n",
    "    # introduce expression values into score columnn\n",
    "    for c in cond_of_interest:\n",
    "        dfRegs = dfRegs.merge(DfTpmAveraged[f'{c} Mean TPM'], left_on=3, right_index=True)\n",
    "        dfRegs = dfRegs[[0, 1, 2, 3, f'{c} Mean TPM', 5]].sort_values(by=[0, 1], ascending=True)\n",
    "        dfRegs.to_csv(f'{out_regions}/{fname}_up{upW}_down{downW}_cond-{c}.bed.gz', sep='\\t', index=False, header=None, quoting=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For each condition merge UP and CONTROL\n",
    "up_files = glob(f'{out_regions}/*_UP_*')\n",
    "ctrl_files = glob(f'{out_regions}/*_CONTROL_*')\n",
    "for c in cond_of_interest:\n",
    "    up = pbt.BedTool([f for f in up_files if c in f][0]).sort()\n",
    "    ctrl = pbt.BedTool([f for f in ctrl_files if c in f][0]).sort()\n",
    "    concat = up.cat(ctrl, postmerge=False).sort()\n",
    "    concat.saveas(f'{out_regions}/terminal_pa_signals_UP+CONTROL_up{upW}_down{downW}_cond-{c}.bed.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DOWN\n",
      "UP\n",
      "CONTROL\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>6</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>4843404</td>\n",
       "      <td>4843655</td>\n",
       "      <td>CONTROL</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000033845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>4967906</td>\n",
       "      <td>4968157</td>\n",
       "      <td>DOWN</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUSG00000033813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9617456</td>\n",
       "      <td>9617707</td>\n",
       "      <td>DOWN</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUSG00000061024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10095023</td>\n",
       "      <td>10095274</td>\n",
       "      <td>CONTROL</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000025917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10206771</td>\n",
       "      <td>10207022</td>\n",
       "      <td>CONTROL</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUSG00000056763</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0         1         2        6  4  5                   3\n",
       "0  chr1   4843404   4843655  CONTROL  0  -  ENSMUSG00000033845\n",
       "0  chr1   4967906   4968157     DOWN  0  +  ENSMUSG00000033813\n",
       "1  chr1   9617456   9617707     DOWN  0  +  ENSMUSG00000061024\n",
       "1  chr1  10095023  10095274  CONTROL  0  -  ENSMUSG00000025917\n",
       "2  chr1  10206771  10207022  CONTROL  0  +  ENSMUSG00000056763"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Save a file with all regions that are included in the metaprofiles\n",
    "allregs = [f for f in glob(f'{out_regions}/terminal*{cond_of_interest[0]}*') if '+' not in f]\n",
    "dfList = []\n",
    "for f in allregs:\n",
    "    group = f.split('/')[-1].split('_')[3]\n",
    "    print(group)\n",
    "    df = readBed(f)\n",
    "    df[4] = 0\n",
    "    df[6] = group\n",
    "    dfList.append(df[[0, 1, 2, 6, 4, 5, 3]])\n",
    "dfCombinedRegs = pd.concat(dfList).sort_values(by=[0, 1])\n",
    "dfCombinedRegs.to_csv(f'{out_combined_regions}/AllRegionsInMetaprofile.bed.gz', sep='\\t', index=False, header=None, quoting=None)\n",
    "dfCombinedRegs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6\n",
       "CONTROL    1739\n",
       "DOWN        831\n",
       "UP          633\n",
       "Name: 0, dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfCombinedRegs.groupby(6).count()[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.7.12 ('metaprofile')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "699bc7c374119b6474c58e324cec7aeb65ec4259d3acfe50fdbc593df4b13d7a"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
