{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from glob import glob\n",
    "import pyranges as pr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import relevant quantfiles\n",
    "quantfiles = sorted(glob('../../data/MihaDeseq/salmon_quantfiles/KO*FCL*') + glob('../../data/MihaDeseq/salmon_quantfiles/*S200*FCL*'))\n",
    "quantfiles = [f for f in quantfiles if \"bulk\" not in f]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Annotation\n",
    "gtf = pr.read_gtf('../../data/genomes/GRCm38/gencode.vM22.primary_assembly.annotation.gtf', as_df=True)\n",
    "gtf = gtf.loc[gtf.Feature == 'transcript']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/klara/anaconda3/envs/JupyterAnalyses/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \n",
      "/home/klara/anaconda3/envs/JupyterAnalyses/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  This is separate from the ipykernel package so we can avoid doing imports until\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>stable_gene_id</th>\n",
       "      <th>transcript_stable_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000102693.1</td>\n",
       "      <td>ENSMUST00000193812.1</td>\n",
       "      <td>ENSMUSG00000102693</td>\n",
       "      <td>ENSMUST00000193812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000064842.1</td>\n",
       "      <td>ENSMUST00000082908.1</td>\n",
       "      <td>ENSMUSG00000064842</td>\n",
       "      <td>ENSMUST00000082908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSMUSG00000051951.5</td>\n",
       "      <td>ENSMUST00000162897.1</td>\n",
       "      <td>ENSMUSG00000051951</td>\n",
       "      <td>ENSMUST00000162897</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENSMUSG00000051951.5</td>\n",
       "      <td>ENSMUST00000159265.1</td>\n",
       "      <td>ENSMUSG00000051951</td>\n",
       "      <td>ENSMUST00000159265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENSMUSG00000051951.5</td>\n",
       "      <td>ENSMUST00000070533.4</td>\n",
       "      <td>ENSMUSG00000051951</td>\n",
       "      <td>ENSMUST00000070533</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 gene_id         transcript_id      stable_gene_id  \\\n",
       "1   ENSMUSG00000102693.1  ENSMUST00000193812.1  ENSMUSG00000102693   \n",
       "4   ENSMUSG00000064842.1  ENSMUST00000082908.1  ENSMUSG00000064842   \n",
       "7   ENSMUSG00000051951.5  ENSMUST00000162897.1  ENSMUSG00000051951   \n",
       "10  ENSMUSG00000051951.5  ENSMUST00000159265.1  ENSMUSG00000051951   \n",
       "13  ENSMUSG00000051951.5  ENSMUST00000070533.4  ENSMUSG00000051951   \n",
       "\n",
       "   transcript_stable_id  \n",
       "1    ENSMUST00000193812  \n",
       "4    ENSMUST00000082908  \n",
       "7    ENSMUST00000162897  \n",
       "10   ENSMUST00000159265  \n",
       "13   ENSMUST00000070533  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TranscriptIds = gtf[['gene_id', 'transcript_id']]\n",
    "TranscriptIds['stable_gene_id'] = TranscriptIds.gene_id.apply(lambda x: x.split('.')[0])\n",
    "TranscriptIds['transcript_stable_id'] = TranscriptIds.transcript_id.apply(lambda x: x.split('.')[0])\n",
    "TranscriptIds.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "TranscriptIds[['transcript_stable_id', 'stable_gene_id']].to_csv('transcript_to_gene_mapping_mm10_V22.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['../../data/MihaDeseq/salmon_quantfiles/KO_E12_FCL_1.quant.sf',\n",
       " '../../data/MihaDeseq/salmon_quantfiles/KO_E12_FCL_2.quant.sf',\n",
       " '../../data/MihaDeseq/salmon_quantfiles/KO_E12_FCL_3.quant.sf',\n",
       " '../../data/MihaDeseq/salmon_quantfiles/S200A_3_FCL_1.quant.sf',\n",
       " '../../data/MihaDeseq/salmon_quantfiles/S200A_3_FCL_2.quant.sf',\n",
       " '../../data/MihaDeseq/salmon_quantfiles/S200A_3_FCL_3.quant.sf',\n",
       " '../../data/MihaDeseq/salmon_quantfiles/S200A_5_FCL_1.quant.sf',\n",
       " '../../data/MihaDeseq/salmon_quantfiles/S200A_5_FCL_2.quant.sf',\n",
       " '../../data/MihaDeseq/salmon_quantfiles/S200WT_1_FCL_1.quant.sf',\n",
       " '../../data/MihaDeseq/salmon_quantfiles/S200WT_1_FCL_2.quant.sf',\n",
       " '../../data/MihaDeseq/salmon_quantfiles/S200WT_2_FCL_3.quant.sf']"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "quantfiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "DfCounts = pd.DataFrame()\n",
    "for i, f in enumerate(quantfiles):\n",
    "    s = f.split('/')[-1].split('.')[0]\n",
    "    df = pd.read_csv(f, sep='\\t', index_col=0).rename(columns={'NumReads': s})\n",
    "    df = df[s]\n",
    "    # print(df.head())\n",
    "    DfCounts[s] = df\n",
    "# Convert transcript counts to gene counts\n",
    "DfCounts = DfCounts.merge(TranscriptIds[['transcript_id', 'stable_gene_id']], left_on='Name', right_on='transcript_id', how='inner').set_index('transcript_id')\n",
    "# Aggregate counts for the same gene_id\n",
    "dfGeneCounts = DfCounts.groupby('stable_gene_id').sum().astype(int)\n",
    "dfGeneCounts.to_csv('ReadCounts.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "DfCountsTPM = pd.DataFrame()\n",
    "for i, f in enumerate(quantfiles):\n",
    "    s = f.split('/')[-1].split('.')[0]\n",
    "    df = pd.read_csv(f, sep='\\t', index_col=0).rename(columns={'TPM': s})\n",
    "    df = df[s]\n",
    "    # print(df.head())\n",
    "    DfCountsTPM[s] = df\n",
    "# Convert transcript counts to gene counts\n",
    "DfCountsTPM = DfCountsTPM.merge(TranscriptIds[['transcript_id', 'stable_gene_id']], left_on='Name', right_on='transcript_id', how='inner').set_index('transcript_id')\n",
    "# Aggregate counts for the same gene_id\n",
    "dfGeneCountsTPM = DfCountsTPM.groupby('stable_gene_id').sum().astype(float)\n",
    "dfGeneCountsTPM.to_csv('TPMSum.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>KO_E12_FCL_1</th>\n",
       "      <th>KO_E12_FCL_2</th>\n",
       "      <th>KO_E12_FCL_3</th>\n",
       "      <th>S200A_3_FCL_1</th>\n",
       "      <th>S200A_3_FCL_2</th>\n",
       "      <th>S200A_3_FCL_3</th>\n",
       "      <th>S200A_5_FCL_1</th>\n",
       "      <th>S200A_5_FCL_2</th>\n",
       "      <th>S200WT_1_FCL_1</th>\n",
       "      <th>S200WT_1_FCL_2</th>\n",
       "      <th>S200WT_2_FCL_3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>stable_gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000001</th>\n",
       "      <td>6.530924</td>\n",
       "      <td>9.363520</td>\n",
       "      <td>5.442524</td>\n",
       "      <td>2.005899</td>\n",
       "      <td>2.375329</td>\n",
       "      <td>8.920072</td>\n",
       "      <td>2.403916</td>\n",
       "      <td>9.464212</td>\n",
       "      <td>0.307615</td>\n",
       "      <td>2.139486</td>\n",
       "      <td>2.946704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000003</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000028</th>\n",
       "      <td>7.710276</td>\n",
       "      <td>9.091545</td>\n",
       "      <td>7.334555</td>\n",
       "      <td>3.027887</td>\n",
       "      <td>5.469507</td>\n",
       "      <td>9.357579</td>\n",
       "      <td>4.251148</td>\n",
       "      <td>8.563081</td>\n",
       "      <td>2.505082</td>\n",
       "      <td>3.344192</td>\n",
       "      <td>4.119250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000031</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.556927</td>\n",
       "      <td>0.583990</td>\n",
       "      <td>0.706341</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000037</th>\n",
       "      <td>0.016093</td>\n",
       "      <td>0.057227</td>\n",
       "      <td>0.041397</td>\n",
       "      <td>0.237239</td>\n",
       "      <td>0.064356</td>\n",
       "      <td>0.163073</td>\n",
       "      <td>0.117013</td>\n",
       "      <td>0.108123</td>\n",
       "      <td>0.454818</td>\n",
       "      <td>0.096248</td>\n",
       "      <td>0.196450</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    KO_E12_FCL_1  KO_E12_FCL_2  KO_E12_FCL_3  S200A_3_FCL_1  \\\n",
       "stable_gene_id                                                                \n",
       "ENSMUSG00000000001      6.530924      9.363520      5.442524       2.005899   \n",
       "ENSMUSG00000000003      0.000000      0.000000      0.000000       0.000000   \n",
       "ENSMUSG00000000028      7.710276      9.091545      7.334555       3.027887   \n",
       "ENSMUSG00000000031      0.000000      0.000000      0.000000       1.556927   \n",
       "ENSMUSG00000000037      0.016093      0.057227      0.041397       0.237239   \n",
       "\n",
       "                    S200A_3_FCL_2  S200A_3_FCL_3  S200A_5_FCL_1  \\\n",
       "stable_gene_id                                                    \n",
       "ENSMUSG00000000001       2.375329       8.920072       2.403916   \n",
       "ENSMUSG00000000003       0.000000       0.000000       0.000000   \n",
       "ENSMUSG00000000028       5.469507       9.357579       4.251148   \n",
       "ENSMUSG00000000031       0.583990       0.706341       0.000000   \n",
       "ENSMUSG00000000037       0.064356       0.163073       0.117013   \n",
       "\n",
       "                    S200A_5_FCL_2  S200WT_1_FCL_1  S200WT_1_FCL_2  \\\n",
       "stable_gene_id                                                      \n",
       "ENSMUSG00000000001       9.464212        0.307615        2.139486   \n",
       "ENSMUSG00000000003       0.000000        0.000000        0.000000   \n",
       "ENSMUSG00000000028       8.563081        2.505082        3.344192   \n",
       "ENSMUSG00000000031       0.000000        0.000000        0.000000   \n",
       "ENSMUSG00000000037       0.108123        0.454818        0.096248   \n",
       "\n",
       "                    S200WT_2_FCL_3  \n",
       "stable_gene_id                      \n",
       "ENSMUSG00000000001        2.946704  \n",
       "ENSMUSG00000000003        0.000000  \n",
       "ENSMUSG00000000028        4.119250  \n",
       "ENSMUSG00000000031        0.000000  \n",
       "ENSMUSG00000000037        0.196450  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfGeneCountsTPM.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>genotype</th>\n",
       "      <th>condition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KO_E12_FCL_1</td>\n",
       "      <td>KO</td>\n",
       "      <td>FCL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KO_E12_FCL_2</td>\n",
       "      <td>KO</td>\n",
       "      <td>FCL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KO_E12_FCL_3</td>\n",
       "      <td>KO</td>\n",
       "      <td>FCL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>S200A_3_FCL_1</td>\n",
       "      <td>S200A</td>\n",
       "      <td>FCL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>S200A_3_FCL_2</td>\n",
       "      <td>S200A</td>\n",
       "      <td>FCL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>S200A_3_FCL_3</td>\n",
       "      <td>S200A</td>\n",
       "      <td>FCL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>S200A_5_FCL_1</td>\n",
       "      <td>S200A</td>\n",
       "      <td>FCL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>S200A_5_FCL_2</td>\n",
       "      <td>S200A</td>\n",
       "      <td>FCL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>S200WT_1_FCL_1</td>\n",
       "      <td>S200WT</td>\n",
       "      <td>FCL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>S200WT_1_FCL_2</td>\n",
       "      <td>S200WT</td>\n",
       "      <td>FCL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>S200WT_2_FCL_3</td>\n",
       "      <td>S200WT</td>\n",
       "      <td>FCL</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            sample genotype condition\n",
       "0     KO_E12_FCL_1       KO       FCL\n",
       "1     KO_E12_FCL_2       KO       FCL\n",
       "2     KO_E12_FCL_3       KO       FCL\n",
       "3    S200A_3_FCL_1    S200A       FCL\n",
       "4    S200A_3_FCL_2    S200A       FCL\n",
       "5    S200A_3_FCL_3    S200A       FCL\n",
       "6    S200A_5_FCL_1    S200A       FCL\n",
       "7    S200A_5_FCL_2    S200A       FCL\n",
       "8   S200WT_1_FCL_1   S200WT       FCL\n",
       "9   S200WT_1_FCL_2   S200WT       FCL\n",
       "10  S200WT_2_FCL_3   S200WT       FCL"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rows = []\n",
    "for c in dfGeneCounts.columns:\n",
    "    genotype = c.split('_')[0]\n",
    "    condition = c.split('_')[2]\n",
    "    row = [c, genotype, condition]\n",
    "    rows.append(row)\n",
    "DfMeta = pd.DataFrame(rows, columns=['sample', 'genotype', 'condition'])\n",
    "DfMeta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfGeneCounts.to_csv('ReadCounts.csv')\n",
    "DfMeta.to_csv('Metadata.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.7.12 ('JupyterAnalyses')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "e3a5e41094ab1c702a06c69f12f608a168aa4760d2953e0e9de1278519583907"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
