{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pybedtools as pbt\n",
    "import pandas as pd\n",
    "from glob import glob\n",
    "import pyranges as pr\n",
    "import os\n",
    "import sys\n",
    "sys.path.append(os.path.abspath('../../../src/'))\n",
    "import GtfBedLibrary as gbl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def readBed(f):\n",
    "    return pd.read_csv(f, sep='\\t', header=None)\n",
    "\n",
    "def getFeatureCenters(df):\n",
    "    \"\"\"\n",
    "    Gets centers of input features, returns bed6.\n",
    "    \"\"\"\n",
    "    df['length'] = df[2] - df[1]\n",
    "    df['middle'] = df['length'].apply(lambda x: int(x / 2) if x % 2 == 0 else int(x / 2) + 1)\n",
    "    # -1 is necessary because start pos is already included\n",
    "    df.loc[df[5]=='+', 'adj_start'] = df.loc[df[5]=='+', 1] + df.loc[df[5]=='+', 'middle'] - 1\n",
    "    df.loc[df[5]=='-', 'adj_start'] = df.loc[df[5]=='-', 2] - df.loc[df[5]=='-', 'middle']\n",
    "    df['adj_end'] = df['adj_start'] + 1\n",
    "    df[['adj_start', 'adj_end']] = df[['adj_start', 'adj_end']].astype(int)\n",
    "    bt = pbt.BedTool().from_dataframe(df[[0, 'adj_start', 'adj_end', 3, 4, 5]]).sort()\n",
    "    return bt\n",
    "\n",
    "def getRegions(bt, fai, window=20):\n",
    "    df = bt.slop(b=window, g=fai).sort().to_dataframe(disable_auto_names=True, header=None)\n",
    "    df['length'] = df[2] - df[1]\n",
    "    # Remove clipped regions that extend beyond chromosome limits\n",
    "    df = df.loc[df.length == df.length.max()]\n",
    "    return df.iloc[:, :-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Files\n",
    "\n",
    "# Params\n",
    "# Region around peaks\n",
    "w = 100\n",
    "\n",
    "# PABP peaks\n",
    "pabpPeaks = glob('../../../data/Pabpc1Pabpc4Iclip_2022/Clippy/*Peaks*')\n",
    "\n",
    "# LIN28A peaks\n",
    "linPeaks = glob('../../../data/LIN28_220626_results/ClippyPeaks/*Peaks*')\n",
    "\n",
    "# full 3'UTRs\n",
    "utrs = pbt.BedTool('../../../data/3UtrAtlas/ThreePrimeUtrsOfMostExpressedTxInS200WT2iL.bed').sort()\n",
    "\n",
    "# fasta index\n",
    "fai = '../../../data/genomes/Goodwright_m39/GRCm39.primary_assembly.genome.fa.fai'\n",
    "\n",
    "# Annotation\n",
    "gtf = pr.read_gtf('../../../data/genomes/Goodwright_m39/gencode.vM28.primary_assembly.annotation.gtf.gz', as_df=True)\n",
    "gtf = gtf.loc[gtf.Feature == 'gene', ['Chromosome', 'Start', 'End', 'gene_id', 'Score', 'Strand']]\n",
    "gtf.gene_id = gtf.gene_id.apply(lambda x: x.split('.')[0])\n",
    "\n",
    "# Expression\n",
    "DfTpm = pd.read_csv('../../../data/MihaDeseq/GeneLevel_TPM_counts.csv')\n",
    "\n",
    "# Save to\n",
    "out_regions = 'ProfileRegions'\n",
    "os.makedirs(out_regions, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BedTool(Shared_LinPeaks.bed.gz)>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge LIN28A peaks and PABP peaks\n",
    "linBts = [pbt.BedTool(f).sort() for f in linPeaks]\n",
    "pabpcBts = [pbt.BedTool(f).sort() for f in pabpPeaks]\n",
    "    # concat\n",
    "catLin = linBts[0].cat(*linBts[1:], postmerge=False).sort()\n",
    "catPabp = pabpcBts[0].cat(*pabpcBts[1:], postmerge=False).sort()\n",
    "    # merge\n",
    "mergedLin = catLin.merge(s=True, d=0, c=[4,5,6], o=['distinct', 'sum', 'distinct']).sort()\n",
    "mergedPabp = catPabp.merge(s=True, d=0, c=[4,5,6], o=['distinct', 'sum', 'distinct']).sort()\n",
    "    # intersect with UTRs\n",
    "mergedLin = mergedLin.intersect(utrs, s=True, wa=True, u=True, nonamecheck=True).sort()\n",
    "mergedPabp = mergedPabp.intersect(utrs, s=True, wa=True, u=True, nonamecheck=True).sort()\n",
    "    # intersect\n",
    "LinShared = mergedLin.intersect(mergedPabp, s=True, wa=True, u=True, nonamecheck=True).sort()\n",
    "LinIndependent = mergedLin.intersect(mergedPabp, s=True, v=True,  nonamecheck=True).sort()\n",
    "    # save peaks\n",
    "LinIndependent.saveas('Independent_LinPeaks.bed.gz')\n",
    "LinShared.saveas('Shared_LinPeaks.bed.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GL456210.1\t9123\t58882\tENSMUSG00000079800\t0\t-\n",
      " GL456210.1\t108389\t110303\tENSMUSG00000095092\t0\t-\n",
      " GL456210.1\t123791\t124928\tENSMUSG00000079192\t0\t+\n",
      " GL456210.1\t135394\t136519\tENSMUSG00000079794\t0\t-\n",
      " GL456210.1\t147791\t149707\tENSMUSG00000094799\t0\t+\n",
      " GL456211.1\t30482\t32396\tENSMUSG00000095250\t0\t-\n",
      " GL456211.1\t54468\t54840\tENSMUSG00000095787\t0\t-\n",
      " GL456211.1\t66954\t67326\tENSMUSG00000096100\t0\t+\n",
      " GL456211.1\t113867\t114758\tENSMUSG00000094054\t0\t+\n",
      " GL456211.1\t160464\t162404\tENSMUSG00000095672\t0\t-\n",
      " "
     ]
    }
   ],
   "source": [
    "# Get gene loci\n",
    "bed_gtf = pbt.BedTool.from_dataframe(gtf.assign(Score=0)).sort()\n",
    "bed_gtf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>KO_2iL Mean TPM</th>\n",
       "      <th>KO_FCL Mean TPM</th>\n",
       "      <th>S200A_2iL Mean TPM</th>\n",
       "      <th>S200A_FCL Mean TPM</th>\n",
       "      <th>S200WT_2iL Mean TPM</th>\n",
       "      <th>S200WT_FCL Mean TPM</th>\n",
       "      <th>WT_2iL Mean TPM</th>\n",
       "      <th>WT_FCL Mean TPM</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>stable_gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000001</th>\n",
       "      <td>7.677950</td>\n",
       "      <td>8.112323</td>\n",
       "      <td>6.880970</td>\n",
       "      <td>6.033886</td>\n",
       "      <td>8.122433</td>\n",
       "      <td>2.797935</td>\n",
       "      <td>4.333492</td>\n",
       "      <td>3.864102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000003</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000028</th>\n",
       "      <td>11.692164</td>\n",
       "      <td>9.045459</td>\n",
       "      <td>10.775650</td>\n",
       "      <td>7.133840</td>\n",
       "      <td>14.348023</td>\n",
       "      <td>4.322841</td>\n",
       "      <td>15.107630</td>\n",
       "      <td>7.196186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000031</th>\n",
       "      <td>1.093472</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.062915</td>\n",
       "      <td>1.569452</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.178640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000000037</th>\n",
       "      <td>1.144660</td>\n",
       "      <td>1.038239</td>\n",
       "      <td>1.305144</td>\n",
       "      <td>1.137961</td>\n",
       "      <td>1.196149</td>\n",
       "      <td>1.249172</td>\n",
       "      <td>1.057581</td>\n",
       "      <td>1.086607</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    KO_2iL Mean TPM  KO_FCL Mean TPM  S200A_2iL Mean TPM  \\\n",
       "stable_gene_id                                                             \n",
       "ENSMUSG00000000001         7.677950         8.112323            6.880970   \n",
       "ENSMUSG00000000003         1.000000         1.000000            1.000000   \n",
       "ENSMUSG00000000028        11.692164         9.045459           10.775650   \n",
       "ENSMUSG00000000031         1.093472         1.000000            1.062915   \n",
       "ENSMUSG00000000037         1.144660         1.038239            1.305144   \n",
       "\n",
       "                    S200A_FCL Mean TPM  S200WT_2iL Mean TPM  \\\n",
       "stable_gene_id                                                \n",
       "ENSMUSG00000000001            6.033886             8.122433   \n",
       "ENSMUSG00000000003            1.000000             1.000000   \n",
       "ENSMUSG00000000028            7.133840            14.348023   \n",
       "ENSMUSG00000000031            1.569452             1.000000   \n",
       "ENSMUSG00000000037            1.137961             1.196149   \n",
       "\n",
       "                    S200WT_FCL Mean TPM  WT_2iL Mean TPM  WT_FCL Mean TPM  \n",
       "stable_gene_id                                                             \n",
       "ENSMUSG00000000001             2.797935         4.333492         3.864102  \n",
       "ENSMUSG00000000003             1.000000         1.000000         1.000000  \n",
       "ENSMUSG00000000028             4.322841        15.107630         7.196186  \n",
       "ENSMUSG00000000031             1.000000         1.000000         1.178640  \n",
       "ENSMUSG00000000037             1.249172         1.057581         1.086607  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get mean TPMs in each condition for gene-level TPMs\n",
    "DfTpm.set_index('stable_gene_id', drop=True, inplace=True)\n",
    "# To avoid 0-division error assign a +1 to all values\n",
    "DfTpm = DfTpm + 1\n",
    "DfTpm.head()\n",
    "conditions = []\n",
    "colnames = []\n",
    "for c in DfTpm.columns:\n",
    "    cond = '_'.join([c.split('_')[el] for el in [0, 2]])\n",
    "    colnames.append(f'{cond}.{c}')\n",
    "    conditions.append(cond)\n",
    "conditions = sorted(set(conditions))\n",
    "DfTpm.columns = colnames\n",
    "\n",
    "DfTpmAveraged = pd.DataFrame()\n",
    "for cond in conditions:\n",
    "    DfTpmAveraged[f'{cond} Mean TPM'] = DfTpm[[c for c in DfTpm.columns if c.split('.')[0] == cond]].mean(axis='columns')\n",
    "DfTpmAveraged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      0        1        2                   6  S200WT_FCL Mean TPM  5\n",
      "0  chr1  4846537  4846738  ENSMUSG00000033845            15.182613  -\n",
      "1  chr1  4846573  4846774  ENSMUSG00000033845            15.182613  -\n",
      "      0        1        2                   6  S200A_FCL Mean TPM  5\n",
      "0  chr1  4846537  4846738  ENSMUSG00000033845            17.58874  -\n",
      "1  chr1  4846573  4846774  ENSMUSG00000033845            17.58874  -\n",
      "      0        1        2                   6  S200WT_2iL Mean TPM  5\n",
      "0  chr1  4846537  4846738  ENSMUSG00000033845            24.556698  -\n",
      "1  chr1  4846573  4846774  ENSMUSG00000033845            24.556698  -\n",
      "      0        1        2                   6  S200WT_FCL Mean TPM  5\n",
      "7  chr1  5232583  5232784  ENSMUSG00000033793             4.794732  +\n",
      "8  chr1  5232620  5232821  ENSMUSG00000033793             4.794732  +\n",
      "      0        1        2                   6  S200A_FCL Mean TPM  5\n",
      "7  chr1  5232583  5232784  ENSMUSG00000033793            7.241819  +\n",
      "8  chr1  5232620  5232821  ENSMUSG00000033793            7.241819  +\n",
      "      0        1        2                   6  S200WT_2iL Mean TPM  5\n",
      "7  chr1  5232583  5232784  ENSMUSG00000033793            11.991824  +\n",
      "8  chr1  5232620  5232821  ENSMUSG00000033793            11.991824  +\n"
     ]
    }
   ],
   "source": [
    "cond_of_interest = ['S200WT_FCL', 'S200A_FCL', 'S200WT_2iL']\n",
    "\n",
    "for landmarks in ['Independent_LinPeaks.bed.gz', 'Shared_LinPeaks.bed.gz']:\n",
    "    fname = landmarks.split('.')[0]\n",
    "    # Read landmarks into dataframe\n",
    "    df_landmarks = readBed(landmarks)\n",
    "    # Drop duplicated landmarks\n",
    "    df_landmarks = df_landmarks.drop_duplicates()\n",
    "    # Get centers\n",
    "    landmark_sites = getFeatureCenters(df_landmarks)\n",
    "    # Map gene ids to features\n",
    "    landmark_sites = landmark_sites.map(bed_gtf, c=4, o='distinct', s=True, nonamecheck=True)\n",
    "    # Get regions around landmarks of interest\n",
    "    dfRegs = getRegions(landmark_sites, fai, window=w)\n",
    "    for c in cond_of_interest:\n",
    "        dfRegs = dfRegs.merge(DfTpmAveraged[f'{c} Mean TPM'], left_on=6, right_index=True)\n",
    "        dfRegs = dfRegs[[0, 1, 2, 6, f'{c} Mean TPM', 5]].sort_values(by=[0, 1], ascending=True)\n",
    "        print(dfRegs.head(2))\n",
    "        dfRegs.to_csv(f'{out_regions}/{fname}_w{w}_cond-{c}.bed.gz', sep='\\t', index=False, header=None, quoting=None)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.7.12 ('JupyterAnalyses')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "e3a5e41094ab1c702a06c69f12f608a168aa4760d2953e0e9de1278519583907"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
